### ISSUE TITLE
INTERSECTS MATCH does not work

### ISSUE THREAD

**Issue by zhangyonguu** [tidwall/tile38#463]
Hey,

I am currently facing the problem that intersects command search option Match seem to not work at all. 

I have a key name 'polygons' and the SCAN command return:

`127.0.0.1:9851> scan polygons`
`{"ok":true,"objects":[{"id":"1","object":{"type":"Polygon","coordinates":[[[113.91832,22.546208],[113.925733,22.556007],[113.935121,22.562477],[113.949134,22.561745],[113.958403,22.550409],[113.957544,22.535585],[113.955399,22.52393],[113.950891,22.521076],[113.937932,22.518083],[113.929006,22.518896],[113.922826,22.522385],[113.92008,22.52837],[113.91832,22.546208]]]}},{"id":"2","object":{"type":"Polygon","coordinates":[[[114.007326,22.626247],[114.017025,22.631001],[114.027067,22.631001],[114.047237,22.629258],[114.057451,22.617691],[114.056765,22.603429],[114.050757,22.597089],[114.032046,22.594553],[114.008528,22.597407],[114.00046,22.607232],[114.007326,22.626247]]]}},{"id":"3","object":{"type":"Polygon","coordinates":[[[113.823649,22.604697],[113.841673,22.610243],[113.861758,22.614838],[113.874804,22.620067],[113.893343,22.605964],[113.908792,22.588373],[113.912569,22.56856],[113.913942,22.531145],[113.900896,22.52734],[113.858668,22.520362],[113.840128,22.56634],[113.823649,22.604697]]]}}],"count":3,"cursor":0,"elapsed":"150.158µs"}`

and

`127.0.0.1:9851> scan polygons match 3`
`{"ok":true,"objects":[{"id":"3","object":{"type":"Polygon","coordinates":[[[113.823649,22.604697],[113.841673,22.610243],[113.861758,22.614838],[113.874804,22.620067],[113.893343,22.605964],[113.908792,22.588373],[113.912569,22.56856],[113.913942,22.531145],[113.900896,22.52734],[113.858668,22.520362],[113.840128,22.56634],[113.823649,22.604697]]]}}],"count":1,"cursor":0,"elapsed":"86.437µs"}`

It seems that SCAN command works well with MATCH.

But when I run the following INTERSECTS command, things start getting abnormal:

`127.0.0.1:9851> intersects polygons match 2 OBJECT {"type":"Feature","geometry":{"type":"Point","coordinates":[113.876386,22.6083}`
`{"ok":true,"objects":[{"id":"3","object":{"type":"Polygon","coordinates":[[[113.823649,22.604697],[113.841673,22.610243],[113.861758,22.614838],[113.874804,22.620067],[113.893343,22.605964],[113.908792,22.588373],[113.912569,22.56856],[113.913942,22.531145],[113.900896,22.52734],[113.858668,22.520362],[113.840128,22.56634],[113.823649,22.604697]]]}}],"count":1,"cursor":0,"elapsed":"93.048µs"}`

**Obviously**, pattern **2** does not match **3**, the result is the same as the one which doesn't have a MATCH option, the MATCH option seem to not work at all, and I have tried other pattern such as *3*, *abc* and so on, but the result is the same.

Maybe I'm not using Tile38 as I should. But please give me some help, thanks!

**Comment by melbania** [comment]
Hi @zhangyonguu,

I cannot reproduce the problem you describe on my side. I am using Tile38 version 1.17.0 ... which version are you using?

In case it helps, the following is a summary of my results when I tested your issue on my local machine:

For me, the query

```
intersects polygons match 2 OBJECT {"type":"Feature","geometry":{"type":"Point","coordinates":[113.876386,22.6083]}}
```

returns

```json
{"ok":true,"objects":[],"count":0,"cursor":0,"elapsed":"126.384µs"}
```

Also, the query

```
intersects polygons match abc OBJECT {"type":"Feature","geometry":{"type":"Point","coordinates":[113.876386,22.6083]}}
```

returns

```json
{"ok":true,"objects":[],"count":0,"cursor":0,"elapsed":"47.363µs"}
```

These are the expected results since the point object defined in the INTERSECTS queries does not intersect any polygons that would match the pattern '2', and there is no polygon that would match the pattern 'abc' in your 'polygons' collection.

However, like you, when I run the query 

```
intersects polygons match 3 OBJECT {"type":"Feature","geometry":{"type":"Point","coordinates":[113.876386,22.6083]}}
```

it does return

```json
{"ok":true,"objects":[{"id":"3","object":{"type":"Polygon","coordinates":[[[113.823649,22.604697],[113.841673,22.610243],[113.861758,22.614838],[113.874804,22.620067],[113.893343,22.605964],[113.908792,22.588373],[113.912569,22.56856],[113.913942,22.531145],[113.900896,22.52734],[113.858668,22.520362],[113.840128,22.56634],[113.823649,22.604697]]]}}],"count":1,"cursor":0,"elapsed":"82.049µs"}
```

as you describe. This is the correct result for this version of the query, because the point defined does intersect the polygon that matches the pattern '3'.

**Comment by zhangyonguu** [comment]
hi, @melbania ,

I am using Tile38 version 1.14.4.

 And I tried it again in version 1.12.3, the result is correct, it seems that the issue occurs only in certain version, maybe I need to upgrade my Tile38 version! 

I still wonder that is it really a bug in version 1.14.4,  please reply to me if you have a conclusion.

Anyway, thanks!

**Comment by melbania** [comment]
Hi @zhangyonguu ,

Yes, I was able to reproduce the bug in version 1.14.4. I got the same wrong results that you describe.

I did some digging and testing with different versions, and it looks like the regression occurred with release 1.14.0, which involved significant changes to the code base. It was fixed with release 1.16.1, which addressed a reporting problem with the MATCH option when used with the NEARBY command (see issue #421).  It looks like the MATCH option was being ignored generally in those interim releases. Now it is back to working again, yay!

Hope that helps!
